File: exercise/php/day23/src/Core/Control/ReindeerPowerUnit.php
<?php

namespace Core\Control;

use External\Deer\Reindeer;

class ReindeerPowerUnit
{
    private Reindeer $reindeer;
    private MagicPowerAmplifier $amplifier;

    public function __construct(Reindeer $reindeer)
    {
        $this->reindeer = $reindeer;
        $this->amplifier = new MagicPowerAmplifier(AmplifierType::BASIC);
    }

    public function harnessMagicPower(): float
    {
        if (!$this->reindeer->needsRest()) {
            $this->reindeer->increaseHarnessingTimes();
            return $this->amplifier->amplify($this->reindeer->getMagicPower());
        }

        return 0;
    }

    public function getReindeer(): Reindeer
    {
        return $this->reindeer;
    }
}

File: exercise/kotlin/day23/src/main/kotlin/core/control/ReindeerPowerUnit.kt
package core.control

import external.deer.Reindeer

class ReindeerPowerUnit(val reindeer: Reindeer) {
    private val amplifier = MagicPowerAmplifier(AmplifierType.BASIC)

    fun harnessMagicPower(): Float {
        return if (!reindeer.needsRest()) {
            reindeer.timesHarnessing++
            amplifier.amplify(reindeer.magicPower)
        } else {
            0f
        }
    }

    fun checkMagicPower() = reindeer.magicPower
}
